File: comms/native/services/networks/SegmentTraceNode.js
import geomUtils from '../base/GeomUtils';
import PinTraceNode from './PinTraceNode';

class SegmentTraceNode extends PinTraceNode {
    static {
        /**
         * A cable segment node in a fiber network trace result
         *
         * Also used to model a list of connected segments (see mutation code)
         */

        this.prototype.type = 'segment';
    }

    /**
     * Init slots of self
     */
    constructor(
        feature,
        tech,
        pins,
        direction,
        dist,
        ewlDist,
        parent = undefined,
        loc = undefined
    ) {
        super(feature, tech, pins, direction, dist, /*parent=*/ parent);

        // Determine if this is a continuation node
        this.leaving =
            parent && parent.type == 'segment' && parent.feature.getUrn() == feature.getUrn();

        // Set trace direction relative to geom (if meaningful)
        this.seg_forward = true; // ENH: Remove need for this
        if (this.leaving) {
            this.seg_forward = pins.side == 'out';
        }

        // Init list of consolidated segments (see mutation)
        this.segments = undefined;

        this.node_id += `-${pins.spec}`; // Used for detecting 'already visited'
        this._loc = loc;
        this._ewlDist = ewlDist;
    }

    //@property
    /**
     * True if this is first node on segment
     */
    _entering() {
        return !this.leaving;
    }

    //@property
    /**
     * For display in trace result
     */
    _fibers() {
        return `${this.pins.rangeSpec()}`;
    }

    //@property
    /**
     * Position on self's feature at which self starts (if partial link)
     */
    _start_coord() {
        if (!this.segments) {
            return undefined;
        }

        const geom = this.segments[0].geometry;

        if (this.seg_forward) {
            return geom.firstCoord();
        } else {
            return geom.lastCoord();
        }
    }

    //@property
    /**
     * Position on self's feature at which self ends (if partial link)
     */
    _stop_coord() {
        // Case: Root node of trace from segment
        if (this.dist == 0.0) {
            return this.start_coord;
        }

        // Case: Unconsolidated segment
        if (!this.segments) {
            return undefined;
        }

        const geom = this.segments[this.segments.length - 1].geometry;

        // Case: Full link
        if (!this.partial) {
            if (this.seg_forward) {
                return geom.lastCoord();
            } else {
                return geom.firstCoord();
            }
        }

        // Case: Zero length segment (prevents problems later)
        if (geom.length() == 0.0) {
            return geom.firstCoord();
        }

        // Find position of stop point along last segment (as proportion of total length)
        // Remember: dist may have been computed from a stored length value
        let pos = (this.dist - this.prev_seg_dist) / (this.full_dist - this.prev_seg_dist);
        if (!this.seg_forward) {
            pos = 1.0 - pos;
        }

        // Compute coordinate at that position
        return geomUtils.coordAtPos(geom, pos);
    }

    /**
     * The coordinates of self's path
     *
     * Returns a list of coords
     */
    coords() {
        // Assumes self is unconsolidated

        if (this.leaving) {
            return [];
        }

        const geom = this.feature.geometry;
        let coords = geom.coordinates;

        if (this.seg_forward) {
            coords = [...coords].reverse();
        }

        return coords;
    }

    /**
     * Convert self to a cable node
     *
     * Returns self
     */
    async convertToCable() {
        if (this.segments === undefined) {
            this.segments = [this.feature];
            this.feature = await this.feature.followRef('cable');
        }

        return this;
    }

    /**
     * Merge self's 'obvious' children into self (to simplify trace)
     */
    consolidate() {
        // Set start distance
        if (this.parent) {
            this.prev_seg_dist = this.parent.dist;
        } else {
            this.prev_seg_dist = 0;
        }

        // While child node relates to same cable .. consolidate it into self
        while (
            this.children.length == 1 &&
            this.children[0].feature.getUrn() == this.feature.getUrn()
        ) {
            const child_node = this.children[0];

            if (child_node.leaving) {
                this.prev_seg_dist = this.dist;
                this.pins.side = child_node.pins.side;
                this.seg_forward = child_node.seg_forward;
                this.dist = child_node.dist;
                this.partial = child_node.partial;
                this.full_dist = child_node.full_dist;
                this.segments = [...this.segments, ...child_node.segments];
            }
            this.children = child_node.children;
        }

        // Consolidate connections
        super.consolidate();

        return this;
    }

    /**
     * The pin on self's feature that is connected to CHILD_PIN of CHILD_NODE
     */
    pinFor(child_node, child_pin) {
        return child_pin;
    }
}

Object.defineProperty(SegmentTraceNode.prototype, 'entering', {
    get() {
        return this._entering();
    }
});

Object.defineProperty(SegmentTraceNode.prototype, 'fibers', {
    get() {
        return this._fibers();
    }
});

Object.defineProperty(SegmentTraceNode.prototype, 'start_coord', {
    get() {
        return this._start_coord();
    }
});

Object.defineProperty(SegmentTraceNode.prototype, 'stop_coord', {
    get() {
        return this._stop_coord();
    }
});

Object.defineProperty(SegmentTraceNode.prototype, 'loc', {
    get() {
        return this._loc;
    }
});

Object.defineProperty(SegmentTraceNode.prototype, 'ewlDist', {
    get() {
        return this._ewlDist;
    }
});

export default SegmentTraceNode;

File: comms/server/networks/segment_trace_node.py
from copy import copy
from .pin_trace_node import PinTraceNode

class SegmentTraceNode(PinTraceNode):
    """
    A cable segment node in a fiber network trace result

    Also used to model a list of connected segments (see mutation code)"""

    type = "segment"

    def __init__(
        self,
        feature,
        tech,
        pins,
        direction,
        dist,
        ewl_dist,
        parent=None,
        loc=None,
    ):
        """
        Init slots of self
        """

        super().__init__(feature, tech, pins, direction, dist, parent=parent)

        # Determine if this is a continuation node
        self.leaving = (
            parent and (parent.type == "segment") and (parent.feature._urn() == feature._urn())
        )

        # Set trace direction relative to geom (if meaningful)
        self.seg_forward = True  # ENH: Remove need for this
        if self.leaving:
            self.seg_forward = pins.side == "out"

        # Init list of consolidated segments (see mutation)
        self.segments = None

        self.node_id += "-{}".format(pins.spec)  # Used for detecting 'already visited'
        self._loc = loc
        self._ewl_dist = ewl_dist

    @property
    def entering(self):
        """
        True if this is first node on segment
        """

        return not self.leaving

    @property
    def fibers(self):
        """
        For display in trace result
        """

        return "{}".format(self.pins.rangeSpec())

    @property
    def loc(self):
        return self._loc

    @property
    def ewl_dist(self):
        return self._ewl_dist

    @property
    def start_coord(self):
        """
        Position on self's feature at which self starts (if partial link)
        """

        if not self.segments:
            return None

        geom = self.segments[0]._primary_geom_field.geom()

        if self.seg_forward:
            return geom.coords[0]
        else:
            return geom.coords[-1]

    @property
    def stop_coord(self):
        """
        Position on self's feature at which self ends (if partial link)
        """

        # Case: Root node of trace from segment
        if self.dist == 0.0:
            return self.start_coord

        # Case: Unconsolidated segment
        if not self.segments:
            return None

        geom = self.segments[-1]._primary_geom_field.geom()

        # Case: Full link
        if not self.partial:
            if self.seg_forward:
                return geom.coords[-1]
            else:
                return geom.coords[0]

        # Case: Zero length segment (prevents problems later)
        if geom.geoLength() == 0.0:
            return geom.coords[0]

        # Find position of stop point along last segment (as proportion of total length)
        # Remember: dist may have been computed from a stored length value
        pos = (self.dist - self.prev_seg_dist) / (self.full_dist - self.prev_seg_dist)
        if not self.seg_forward:
            pos = 1.0 - pos

        # Compute coordinate at that position
        return geom.geoCoordAtPos(pos)

    def coords(self):
        """
        The coordinates of self's path

        Returns a list of coords"""

        # Assumes self is unconsolidated

        if self.leaving:
            return []

        geom = self.feature._primary_geom_field.geom()
        coords = geom.coords

        if self.forward:
            coords = coords[::-1]

        return coords

    def convertToCable(self):
        """
        Convert self to a cable node

        Returns self"""

        if self.segments is None:
            self.segments = [self.feature]
            self.feature = self.feature._field("cable").rec()

        return self

    def consolidate(self):
        """
        Merge self's 'obvious' children into self (to simplify trace)
        """

        # Set start distance
        if self.parent:
            self.prev_seg_dist = self.parent.dist
        else:
            self.prev_seg_dist = 0

        # While child node relates to same cable .. consolidate it into self
        while len(self.children) == 1 and (self.children[0].feature == self.feature):
            child_node = self.children[0]

            if child_node.leaving:
                self.prev_seg_dist = self.dist
                # Copy to ensure that we don't update shared pin range
                self.pins = copy(self.pins)
                self.pins.side = child_node.pins.side
                self.seg_forward = child_node.seg_forward
                self.dist = child_node.dist
                self.partial = child_node.partial
                self.full_dist = child_node.full_dist
                self.segments += child_node.segments
            self.children = child_node.children

        # Consolidate connections
        super().consolidate()

        return self

    def pinFor(self, child_node, child_pin):
        """
        The pin on self's feature that is connected to CHILD_PIN of CHILD_NODE
        """

        return child_pin
